dds/DCPS/XTypes/DynamicDataImpl.h

```diff
@@ -61,524 +45,898 @@ class OpenDDS_Dcps_Export DynamicDataImpl : public virtual DCPS::LocalObject<DDS
+      size_t& mutable_running_total) const;
+    bool serialize_discriminator_member_xcdr2(DCPS::Serializer& ser, CORBA::Long value,
+      const DDS::DynamicType_var& disc_type, DDS::ExtensibilityKind extensibility) const;
+    bool serialized_size_selected_member_xcdr2(const DCPS::Encoding& encoding, size_t& size,
+      DDS::MemberId selected_id, DDS::ExtensibilityKind extensibility,
+      size_t& mutable_running_total) const;
+    bool serialize_selected_member_xcdr2(DCPS::Serializer& ser, DDS::MemberId selected_id,
+                                         DDS::ExtensibilityKind extensibility) const;
+    bool select_union_member(CORBA::Long disc_value, bool& found_selected_member,
+                             DDS::MemberDescriptor_var& selected_md) const;
+    bool serialized_size_union_xcdr2(const DCPS::Encoding& encoding, size_t& size) const;
+    bool serialize_union_xcdr2(DCPS::Serializer& ser) const;
+    bool serialized_size_union_xcdr1(const DCPS::Encoding& encoding, size_t& size) const;
+    bool serialize_union_xcdr1(DCPS::Serializer& ser) const;
+    bool serialized_size_union(const DCPS::Encoding& encoding, size_t& size) const;
+    bool serialize_union(DCPS::Serializer& ser) const;
+
+    // Internal data
+    OPENDDS_MAP(DDS::MemberId, SingleValue) single_map_;
+    OPENDDS_MAP(DDS::MemberId, SequenceValue) sequence_map_;
+    OPENDDS_MAP(DDS::MemberId, DDS::DynamicData_var) complex_map_;
+
+    const DDS::DynamicType_var& type_;
+    const DynamicDataImpl* data_;
+  };
 
-  /// The alignment state that a Serializer object associated with this DynamicData
-  /// object will be set to.
-  DCPS::Serializer::RdState align_state_;
+  DataContainer container_;
 
-  /// Each public interface creates a new Serializer object with a message block
-  /// chain that is a duplicate from chain_.
-  DCPS::Serializer strm_;
+  friend void serialized_size(const DCPS::Encoding& encoding, size_t& size, const DynamicDataImpl& data);
+  friend bool operator<<(DCPS::Serializer& ser, const DynamicDataImpl& data);
+};
 
-  /// Message block chains created during each get_*_value or get_*_values method's
-  /// execution that need to be released when the method ends. Those chains are created
-  /// when the method skips a nested aggregated type (i.e., struct and union) by
-  /// calling skip_aggregated_member().
-  IntermediateChains chains_to_release;
+} // namespace XTypes
 
-  /// This DynamicData object holds data for this type.
-  DDS::DynamicType_var type_;
+namespace DCPS {
 
-  static const ACE_CDR::ULong ITEM_COUNT_INVALID = ACE_UINT32_MAX;
+OpenDDS_Dcps_Export
+void serialized_size(const Encoding& encoding, size_t& size, const XTypes::DynamicDataImpl& data);
 
-  /// Cache the number of items (i.e., members or elements) in the data it holds.
-  ACE_CDR::ULong item_count_;
-};
+OpenDDS_Dcps_Export
+bool operator<<(Serializer& ser, const XTypes::DynamicDataImpl& data);
```

Comment:
I still need these to take `DynamicDataImpl` because they are called internally in `DynamicDataImpl` class. So maybe we add another pair of `operator<<` and `serialized_size` which take `DDS::DynamicData` and delegate the serialization operations to the ones that take `DynamicDataImpl` when it's the specific DynamicData implementation being used?